src/core/MIGRATION.md

```diff
@@ -0,0 +1,681 @@
+
+```ts
+import { PluginInitializerContext, CoreSetup, PluginStop } from '../../../core/public';
+
+export class Plugin {
+  constructor(initializerContext: PluginInitializerContext) {
+  }
+
+  public setup(core: CoreSetup) {
+    // called when plugin is setting up
+  }
+
+  public stop(core: PluginStop) {
+    // called when plugin is torn down, aka window.onbeforeunload
+  }
+}
+```
+
+**[4] `server/index.ts`** is the entry-point into the server-side code of this plugin. It is identical in almost every way to the client-side entry-point:
+
+```ts
+import { PluginInitializerContext } from '../../../core/server';
+import { Plugin } from './plugin';
+
+export function plugin(initializerContext: PluginInitializerContext) {
+  return new Plugin(initializerContext);
+}
+```
+
+**[5] `server/plugin.ts`** is the server-side plugin definition. The _shape_ of this plugin is the same as it's client-side counter-part:
+
+```ts
+import { PluginInitializerContext, CoreSetup, PluginStop } from '../../../core/server';
+
+export class Plugin {
+  constructor(initializerContext: PluginInitializerContext) {
+  }
+
+  public setup(core: CoreSetup) {
+    // called when plugin is setting up during Kibana's startup sequence
+  }
+
+  public stop(core: PluginStop) {
+    // called when plugin is torn down during Kibana's shutdown sequence
+  }
+}
+```
+
+The platform does not impose any technical restrictions on how the internals of the plugin are architected, though there are certain considerations related to how plugins interact with core and how plugins interact with other plugins that may greatly impact how they are built.
+
+### Services
+
+The various independent domains that make up `core` are represented by a series of services, and many of those services expose public interfaces that are provided to _all_ plugins via the first argument of their `setup` and `stop` functions. The interface varies from service to service, but it is always accessed through this argument.
+
+For example, the core `UiSettings` service exposes a function `get` to all plugin `setup` functions. To use this function to retrieve a specific UI setting, a plugin just accesses it off of the first argument:
+
+```ts
+import { CoreSetup } from '../../../core/public';
+
+export class Plugin {
+  public setup(core: CoreSetup) {
+    core.uiSettings.get('courier:maxShardsBeforeCryTime');
+  }
+}
+```
+
+Different service interfaces can and will be passed to `setup` and `stop` because certain functionality makes sense in the context of a running plugin while other types of functionality may have restrictions or may only make sense in the context of a plugin that is stopping.
+
+For example, the `stop` function in the browser gets invoked as part of the `window.onbeforeunload` event, which means you can't necessarily execute asynchronous code here in a reliable way. For that reason, `core` likely wouldn't provide any asynchronous functions to plugin `stop` functions in the browser.
```

Comment:
imho it makes sense to start Plugin introduction from lifecycles and describe their signature. Don't see any `start` mentions in the doc

Reply:
To clarify, do you mean that you'd like to have a description in this document of the three lifecycle functions and how lifecycle functions work in general?

Reply:
- list of lifecycles
- their meaning 
- their signature
- their calling order

Reply:
I agree that this would be valuable context. It would also be good to include the plugin factory function in the lifecycle calling order.

Reply:
I pushed more information about lifecycles with a focus on relevant information for the sake of migration 5d343de2fcc4452470bc7bd3e6d9b9e51e459b13